cjdns-bytes/src/utils.rs

```diff
@@ -1,150 +1,224 @@
+            Ok(u64::from_be_bytes(bytes_array))
+        }
 
-    pub fn read_all_mut(&mut self) -> &[u8] {
-        self.take_bytes(self.len()).expect("attempting to read data more than slice have")
-    }
+        pub fn read_array_32(&mut self) -> Result<[u8; 32]> {
+            let bytes_32_slice = self.read_slice(32)?;
+            let bytes_32_array = <[u8; 32]>::try_from(bytes_32_slice).expect("invalid slice size");
+            Ok(bytes_32_array)
+        }
 
-    pub fn take_bytes(&mut self, count: usize) -> Result<&[u8]> {
-        if self.len() < count {
-            return Err(());
+        pub fn pick_remainder(&self) -> &'a [u8] {
+            self.0
+        }
+
+        pub fn read_remainder(&mut self) -> &'a [u8] {
+            self.read_slice(self.len()).expect("attempting to read data more than slice have")
+        }
+
+        pub fn read_slice(&mut self, count: usize) -> Result<&'a [u8]> {
+            if self.len() < count {
+                return Err(InsufficientBuffer);
+            }
+            let (ret_bytes, rest_bytes) = self.0.split_at(count);
+            self.0 = rest_bytes;
+            Ok(ret_bytes)
+        }
+
+        fn len(&self) -> usize {
+            self.0.len()
         }
-        let (ret_bytes, rest_bytes) = self.0.split_at(count);
-        self.0 = rest_bytes;
-        Ok(ret_bytes)
     }
 
-    fn len(&self) -> usize {
-        self.0.len()
+    #[cfg(test)]
+    mod tests {
+        use super::*;
+
+        #[test]
+        fn test_reader() {
+            let bytes = vec![0u8; 64];
+            let mut reader = Reader::new(&bytes);
+
+            let array_32 = reader.read_array_32().expect("bytes are read out");
+            assert_eq!(array_32, [0; 32]);
+            let num64 = reader.read_u64_be().expect("bytes are read out");
+            assert_eq!(num64.to_be_bytes(), [0; 8]);
+            let num32 = reader.read_u32_be().expect("bytes are read out");
+            assert_eq!(num32.to_be_bytes(), [0; 4]);
+            let num16 = reader.read_u16_be().expect("bytes are read out");
+            assert_eq!(num16.to_be_bytes(), [0; 2]);
+            let num8 = reader.read_u8().expect("bytes are read out");
+            assert_eq!(num8, 0);
+            // checking read out of bounds
+            assert!(reader.read_slice(20).is_err());
+            // reading last bytes
+            reader.read_slice(17).expect("bytes are read out");
+            assert_eq!(reader.len(), 0);
+        }
+
+        #[test]
+        fn test_read_all() {
+            let bytes = vec![0; 32];
+            let mut reader = Reader::new(&bytes);
+
+            assert_eq!(bytes.as_slice(), reader.pick_remainder());
+            let _ = reader.read_u16_be();
+            assert_eq!(&bytes[2..], reader.pick_remainder());
+            let _ = reader.read_u16_be();
+            assert_eq!(&bytes[4..], reader.read_remainder());
+            assert_eq!(0, reader.len())
+        }
+
+        #[test]
+        #[should_panic]
```

Comment:
This is a non-working test.
There is can be only one panic, but you are placing four of them in you test. Only the first one will be thrown, others are not tested because that code is unreachable. Need to split into separate tests.

Reply:
done